src/orquestra/sdk/_base/cli/_dorq/_arg_resolvers.py

```diff
@@ -121,18 +167,32 @@ def __init__(
         wf_run_repo=_repos.WorkflowRunRepo(),
         prompter=_prompts.Prompter(),
         presenter=_presenters.PromptPresenter(),
+        spaces_resolver=SpacesResolver(),
     ):
         self._wf_run_repo = wf_run_repo
         self._prompter = prompter
         self._presenter = presenter
+        self._spaces_resolver = spaces_resolver
 
     def resolve_id(
         self, wf_run_id: t.Optional[WorkflowRunId], config: ConfigName
     ) -> WorkflowRunId:
         if wf_run_id is not None:
             return wf_run_id
 
-        wfs = self._wf_run_repo.list_wf_runs(config)
+        try:
+            resolved_workspace_id = self._spaces_resolver.resolve_workspace_id(config)
+            resolved_project_id = self._spaces_resolver.resolve_project_id(
+                config, workspace_id=resolved_workspace_id
+            )
+            project = ProjectRef(
+                workspace_id=resolved_workspace_id, project_id=resolved_project_id
+            )
+        except NotImplementedError:
+            # if run on runtime that doesn't support workspaces
+            project = None
```

Comment:
**nitpick**: `NotImplementedError` is a generic error that can be raised by all sorts of code, including 3rd-party code we don't have insight into. I'm aware that we need to support runtimes that don't make use of projects and workspaces. What exactly is the component that raises `NotImplementedError` today? Can we structure it in a different way, e.g. using a concrete error type for this particular use case?

Reply:
One idea: add `WorkspacesNotSupportedError` class, use it in the `RuntimeInterface`'s contract that implementations should raise it if a caller uses `RuntimeInterface.list_workspaces()` when the runtime impl doesn't support it.